dnf/transaction_sr.py

```diff
@@ -0,0 +1,508 @@
+                    self._base.environment_remove(env_id)
+                else:
+                    errors.append(TransactionError(
+                        _('Unexpected value of environment action "{action}" for environment "{env}".')
+                            .format(action=action, env=env_id)
+                    ))
+            except KeyError as e:
+                errors.append(TransactionError(
+                    _('Missing object key "{key}" in an environment.').format(key=e.args[0])
+                ))
+            except ValueError as e:
+                # env does not exist
+                errors.append(TransactionError(str(e)))
+            except dnf.exceptions.CompsError as e:
+                # installed env not found for upgrade/remove
+                # env not available anymore for upgrade
+                errors.append(TransactionError(str(e)))
+
+        if errors:
+            raise TransactionFileError(fn, errors)
+
+    def post_transaction(self):
+        """
+        Sets reasons in the transaction history to values from the stored transaction.
+
+        Also serves to check whether additional packages were pulled in by the
+        transaction, which results in an error (unless ignore_extras is True).
+        """
+
+        if not self._base.transaction:
+            return
+
+        errors = []
+
+        for tsi in self._base.transaction:
+            try:
+                pkg = tsi.pkg
+            except KeyError as e:
+                # the transaction item has no package, happens for action == "Reason Change"
+                continue
+
+            nevra = str(pkg)
+
+            if nevra not in self._nevra_cache:
+                # if ignore_installed is True, we don't want to check for
+                # Upgraded/Downgraded/Reinstalled extras in the transaction,
+                # basically those may be installed and we are ignoring them
+                if not self._ignore_installed or not tsi.action in (
+                    libdnf.transaction.TransactionItemAction_UPGRADED,
+                    libdnf.transaction.TransactionItemAction_DOWNGRADED,
+                    libdnf.transaction.TransactionItemAction_REINSTALLED
+                ):
+                    msg = _('Rpm nevra "{nevra}", which is not present in the transaction file, was pulled '
+                        'into the transaction.'
+                    ).format(nevra=nevra)
+
+                    if not self._ignore_extras:
+                        errors.append(TransactionError(msg))
+                    else:
+                        self._warnings.append(msg)
+
+            try:
+                replay_reason = self._nevra_reason_cache[nevra]
+
+                if tsi.action in (
+                    libdnf.transaction.TransactionItemAction_INSTALL,
+                    libdnf.transaction.TransactionItemAction_REMOVE
+                ) or libdnf.transaction.TransactionItemReasonCompare(replay_reason, tsi.reason) > 0:
```

Comment:
Not sure here if it will propagate the reason correctly. Let's try install `pkgA` that requires `pkgB`. Then replay the transaction and I suggest that it will result that both will be user installed.

Reply:
This seems to be a basic use case IIUC. There wasn't this exact test case, so I've added one here:
https://github.com/lukash/ci-dnf-stack/blob/52bfdd14f50160c5cf589b613d7cf0d5bb03b594/dnf-behave-tests/features/transaction-sr/replay.feature#L9

And it works correctly (unless I misunderstood). In any case, check the existing tests and I encourage you again to verify any cases you think of using the test suite. If it's not covered, the test case is useful to include in any case and if it's broken, I'll have a reproducer straight away, which will save time.